contracts/OptimisticArbitrator.sol

```diff
@@ -0,0 +1,166 @@
+        require(_getIdentifierWhitelist().isIdentifierSupported(priceIdentifier), "Unsupported identifier");
+        require(_getCollateralWhitelist().isOnWhitelist(address(currency)), "Unsupported currency");
+        require(timestamp <= getCurrentTime(), "Timestamp in future");
+        require(ancillaryData.length <= OO_ANCILLARY_DATA_LIMIT, "Ancillary Data too long");
+
+        requests[requestId] = Request({
+            proposer: address(0),
+            disputer: address(0),
+            currency: currency,
+            settled: false,
+            proposedPrice: 0,
+            reward: reward,
+            finalFee: _getStore().computeFinalFee(address(currency)).rawValue,
+            bond: bond,
+            customLiveness: customLiveness,
+            expirationTime: 0
+        });
+
+        if (reward > 0) currency.safeTransferFrom(msg.sender, address(this), reward);
+    }
+
+    function proposePrice(
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        int256 proposedPrice
+    ) public {
+        Request storage request = requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
+        require(address(request.currency) != address(0), "Price not requested");
+        require(request.proposer == address(0), "Current proposal in liveness");
+
+        request.proposer = msg.sender;
+        request.proposedPrice = proposedPrice;
+        request.expirationTime = uint64(getCurrentTime()) + request.customLiveness;
+
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond + request.finalFee);
+    }
+
+    function disputePrice(uint256 timestamp, bytes memory ancillaryData) public {
+        Request storage request = requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
+        require(request.proposer != address(0), "No proposed price to dispute");
+        require(request.disputer == address(0), "Proposal already disputed");
+        require(uint64(getCurrentTime()) < request.expirationTime, "Proposal past liveness");
+
+        request.disputer = msg.sender;
+
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond + request.finalFee);
+
+        OptimisticOracleV2Interface oo = _getOptimisticOracle();
+
+        request.currency.approve(address(oo), 2 * (request.bond + request.finalFee) + request.reward);
+
+        oo.requestPrice(priceIdentifier, timestamp, ancillaryData, request.currency, request.reward);
+        oo.setBond(priceIdentifier, timestamp, ancillaryData, request.bond);
+        oo.proposePriceFor(
+            request.proposer,
+            address(this),
+            priceIdentifier,
+            timestamp,
+            ancillaryData,
+            request.proposedPrice
+        );
+        oo.disputePriceFor(msg.sender, address(this), priceIdentifier, timestamp, ancillaryData);
+
+        delete requests[_getId(msg.sender, priceIdentifier, timestamp, ancillaryData)];
```

Comment:
I could be reading this contract incorrectly, but I'm not sure this works as intended.

If a user were to go through this flow with the same price identifier, timestamp, and ancillary data, get disputed, and then start over, the OO that this is making requests to would block the second's dispute flow since that has already been requested.

If the user makes the second request with one of those values being different, I don't think we would need this contract since the regular OO would view that as a completely new request.

If the intention is to have the user pass in the same details, but have the OO resolve them as distinct requests, I think this contract would need to internally modify either the identifier, timestamp, or ancillary data on the second request to ensure it doesn't fall into the same bucket as the first in the OO.

Does that make sense?

Reply:
After team discussion, we agreed to eliminate internal price requests.